Objects/longobject.c

```diff
@@ -4565,11 +4595,23 @@ _PyLong_Rshift(PyObject *a, size_t shiftby)
 static PyObject *
 long_lshift1(PyLongObject *a, Py_ssize_t wordshift, digit remshift)
 {
-    /* This version due to Tim Peters */
     PyLongObject *z = NULL;
     Py_ssize_t oldsize, newsize, i, j;
     twodigits accum;
 
+    if (wordshift == 0 && IS_MEDIUM_VALUE(a) &&
+        (a->ob_digit[0] & ~(PyLong_MASK >> remshift)) == 0) {
+        stwodigits sval = (stwodigits)((twodigits)medium_value(a) << remshift);
```

Comment:
I'd suggest splitting into negative and nonnegative cases. It'll be easier to read that way, and a good compiler can probably figure out how to recombine them if that's more efficient for the processor. (And the code as written isn't strictly portable. There are some crazy corner cases in C's integer handling: the left operand of `<<` here doesn't even necessarily have unsigned type, because if `twodigits` is smaller than `int` (which is certainly allowed by the standard) then the left operand gets promoted to an `int`. We're unlikely to run into this case in practice, but we might as well go for the direct and obviously correct spelling anyway.)

Reply:
The checking `a->ob_digit[0] & ~(PyLong_MASK >> remshift) == 0` ensures no bit overflowing or discarding, that is valid bits would stay in the range covered by twodigit.

Besides, I'm sorry but I couldn't catch the idea about "splitting into negative and nonnegative cases" - it's both C's left shifting that applied to "medium value". Could you please explain it further?

Reply:
So the problem is that it takes some mental effort to understand why casting to `twodigits`, shifting, and then casting back to `stwodigits` gives the correct result. Moreover, there are corner cases (admittedly unlikely to occur in practice, but still theoretically possible) where it's still undefined behaviour, if the integer promotions promote an `twodigits` to `int`. See C99 §6.3.1.1, paragraph 2.

Because of those two things, I'd prefer a more direct spelling: if `medium_value(a)` is nonnegative, do `medium_value(a) << remshift` (without any casting). If `medium_value(a)` is negative, do `-((-medium_value(a)) << remshift)` (again without any casting). So something like `stwodigits sval = m < 0 ? -(-m << remshift) : m << remshift`, where `m` is `medium_value(a)`.